src/combine.jl

```diff
@@ -139,3 +139,89 @@ function Base.join{T,N,D,Ax}(As::AxisArray{T,N,D,Ax}...; fillvalue::T=zero(T),
     return result
 
 end #join
+
+function greatest_common_axis(As::AxisArray...)
+    length(As) == 1 && return ndims(first(As))
+
+    for (i, zip_axes) in enumerate(zip(axes.(As)...))
+        if !all(ax -> ax == zip_axes[1], zip_axes[2:end])
+            return i - 1
+        end
+    end
+
+    return minimum(map(ndims, As))
+end
+
+function flatten_array_axes(array_name, array_axes)
+    map(zip(repeated(array_name), product(map(Ax->Ax.val, array_axes)...))) do tup
+        tup_name, tup_idx = tup
+        return (tup_name, tup_idx...)
+    end
+end
+
+function flatten_axes(array_names, array_axes)
+    collect(chain(map(flatten_array_axes, array_names, array_axes)...))
+end
+
+"""
+    flatten(As::AxisArray...) -> AxisArray
+    flatten(last_dim::Integer, As::AxisArray...) -> AxisArray
+
+Concatenates AxisArrays with equal leading axes into a single AxisArray.
+All additional axes in any of the arrays are flattened into a single additional
+CategoricalVector{Tuple} axis.
+
+### Arguments
+
+* `last_dim::Integer`: (optional) the greatest common dimension to share between all input
+                       arrays. The remaining axes are flattened. If this argument is not
+                       provided, the greatest common axis found among the input arrays is
+                       used. All preceeding axes must also be common to each input array, at
+                       the same dimension. Values from 0 up to one more than the minimum
+                       number of dimensions across all input arrays are allowed.
+* `As::AxisArray...`: AxisArrays to be flattened together.
+"""
+function flatten(As::AxisArray...; kwargs...)
+    gca = greatest_common_axis(As...)
+
+    return _flatten(gca, As...; kwargs...)
+end
+
+function flatten(last_dim::Integer, As::AxisArray...; kwargs...)
```

Comment:
The only sad thing here is that the result of `flatten` won't be inferrable. Perhaps you don't care, but if there could be an inferrable API, it might be nicer. If you're just comparing `AxisArrays`, you should be able to split the axes into "common" and "distinct" subsets. For example,
```julia
using AxisArrays

# split out the first axis of each axisarray into a separate tuple
@inline selectcommon(As::AxisArray...) = selectcommon(axes.(As)...)
@inline selectcommon(axs::Tuple...) = _selectcommon(first.(axs), Base.tail.(axs))
# If the next axis has common name, keep it and then keep going
@inline _selectcommon(axes1::NTuple{N, Axis{name}}, trailingaxes::NTuple{N, Any}) where {N,name} = (axes1, _selectcommon(first.(trailingaxes), Base.tail.(trailingaxes))...)
# If they don't have the same Axis name, we drop the rest
@inline _selectcommon(axes1, trailingaxes) = ()

A = AxisArray(rand(3,3,3), :x, :y, :z)
B = AxisArray(rand(3,3,3), :x, :y, :t)

cmn = selectcommon(A, B)
```
should generate a tuple-of-tuples, each element that containing one common axis from each input AxisArray (in this case, it generates `(xtuple, ytuple)` where `xtuple` contains the two `:x` axes and `ytuple` contains the two `:y` axes). You can get the (non-common) trailing axes with a call to `Base.IteratorsMD.split`, just using the `N` in the `NTuple` for the length of `cmn`.